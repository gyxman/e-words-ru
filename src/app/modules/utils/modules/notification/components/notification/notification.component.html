<p>{{ info?.text }}</p>
<p>{{ info?.type }}</p>
