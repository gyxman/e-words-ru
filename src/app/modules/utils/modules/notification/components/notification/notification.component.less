@import '../../../../../../../styles/index';

:host {
    display: block;
    margin: 8px;
}

.wrapper {
    display: flex;
    max-height: 60px;
    box-shadow: 0 2px 8px 0 fade(@color-black, 24%);
    overflow: hidden;

    &_success {
        & .icon-wrapper {
            background-color: @color-green-light;
        }
    }

    &_warning {
        & .icon-wrapper {
            background-color: @color-orange;
        }
    }

    &_error {
        & .icon-wrapper {
            background-color: @color-red;
        }
    }
}

.icon-wrapper {
    position: relative;
    display: flex;
    align-items: center;
    padding-left: 15px;
    margin-right: 30px;
    flex-shrink: 0;
    color: @color-white;

    &:before {
        .clear();
        position: absolute;
        top: 50%;
        right: -20px;
        width: 50px;
        height: 50px;
        background-color: inherit;
        transform: rotate(45deg) translateY(-50%);
        transform-origin: top;
    }
}

.icon {
    position: relative;
    z-index: 2;
    width: 24px;
    height: 24px;
}

.info {
    .font-p2();
    padding: 10px;
}
