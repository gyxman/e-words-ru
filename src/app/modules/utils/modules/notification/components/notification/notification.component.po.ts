import {PageObject} from '../../../../../../testing/page-object';
import {DebugElement} from '@angular/core';

export class NotificationComponentPo<T> extends PageObject<T> {
    get icon(): DebugElement {
        return this.getByAutomationId('icon');
    }
}
