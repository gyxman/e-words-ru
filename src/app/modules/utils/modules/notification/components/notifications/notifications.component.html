<ng-container *ngIf="notificationStream$ | async"></ng-container>
<template #notificationContainer></template>
