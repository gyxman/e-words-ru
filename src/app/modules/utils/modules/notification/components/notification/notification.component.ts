import {ChangeDetectionStrategy, Component, HostBinding, Input} from '@angular/core';
import {NotificationModel} from '../../models/notification';
import {animate, state, style, transition, trigger} from '@angular/animations';
import {MatIconRegistry} from '@angular/material/icon';
import {DomSanitizer} from '@angular/platform-browser';

@Component({
    selector: 'app-notification',
    templateUrl: './notification.component.html',
    styleUrls: ['./notification.component.less'],
    changeDetection: ChangeDetectionStrategy.OnPush,
    animations: [
        trigger('expand', [
            state('in', style({height: '*'})),
            transition(':leave', [
                style({height: '*'}),
                animate(
                    200,
                    style({height: 0, opacity: 0, transform: 'translateX(110%)'}),
                ),
            ]),
            transition(':enter', [
                style({height: 0, opacity: 0, transform: 'translateX(110%)'}),
                animate(
                    200,
                    style({height: '*', opacity: 1, transform: 'translateX(0)'}),
                ),
            ]),
        ]),
    ],
})
export class NotificationComponent {
    @HostBinding('@expand') readonly expand;

    @Input() info: NotificationModel;

    constructor(iconRegistry: MatIconRegistry, sanitizer: DomSanitizer) {
        iconRegistry.addSvgIcon(
            'success',
            sanitizer.bypassSecurityTrustResourceUrl(
                'assets/icons/notifications/icon-success.svg',
            ),
        );
        iconRegistry.addSvgIcon(
            'warning',
            sanitizer.bypassSecurityTrustResourceUrl(
                'assets/icons/notifications/icon-warning.svg',
            ),
        );
        iconRegistry.addSvgIcon(
            'error',
            sanitizer.bypassSecurityTrustResourceUrl(
                'assets/icons/notifications/icon-error.svg',
            ),
        );
    }

    get classes() {
        const {type} = this.info;

        return {
            [`wrapper_${type}`]: true,
        };
    }
}
