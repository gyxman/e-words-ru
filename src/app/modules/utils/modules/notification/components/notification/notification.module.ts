import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {NotificationComponent} from './notification.component';
import {MatIconModule} from '@angular/material/icon';

@NgModule({
    imports: [CommonModule, MatIconModule],
    declarations: [NotificationComponent],
    exports: [NotificationComponent],
})
export class NotificationModule {}
