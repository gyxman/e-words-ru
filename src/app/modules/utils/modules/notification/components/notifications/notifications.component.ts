import {
    ChangeDetectionStrategy,
    Component,
    ComponentFactoryResolver,
    ViewChild,
    ViewContainerRef,
} from '@angular/core';
import {NotificationService} from '../../services/notification.service';
import {NotificationModel} from '../../models/notification';
import {NotificationComponent} from '../notification/notification.component';
import {tap} from 'rxjs/operators';

@Component({
    selector: 'app-notifications',
    templateUrl: './notifications.component.html',
    styleUrls: ['./notifications.component.less'],
    changeDetection: ChangeDetectionStrategy.Default,
})
export class NotificationsComponent {
    @ViewChild('notificationContainer', {read: ViewContainerRef})
    notificationContainer: ViewContainerRef;

    components: {[key: number]: any} = {};
    componentRef: any;

    notificationStream$ = this.notificationService.notificationsStream$.pipe(
        tap(info => {
            this.createComponent(info);
        }),
    );

    constructor(
        private notificationService: NotificationService,
        private resolver: ComponentFactoryResolver,
    ) {}

    private createComponent(notification: NotificationModel) {
        const factory = this.resolver.resolveComponentFactory(NotificationComponent);
        const id = Date.now();

        this.componentRef = this.notificationContainer.createComponent(factory);
        this.componentRef.instance.info = notification;

        this.components[id] = this.componentRef.hostView;

        setTimeout(() => {
            this.removeComponent(id);
        }, notification.time || 3000);
    }

    private removeComponent(index: number) {
        const componentRef = this.components[index];
        const componentIndex: number = this.notificationContainer.indexOf(componentRef);

        this.notificationContainer.remove(componentIndex);
        delete this.components[index];
    }
}
