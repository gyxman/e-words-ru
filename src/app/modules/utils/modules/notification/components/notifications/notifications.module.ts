import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {NotificationModule} from '../notification/notification.module';
import {NotificationsComponent} from './notifications.component';
import {NotificationService} from '../../services/notification.service';

@NgModule({
    imports: [CommonModule, NotificationModule],
    declarations: [NotificationsComponent],
    providers: [NotificationService],
    exports: [NotificationsComponent],
})
export class NotificationsModule {}
