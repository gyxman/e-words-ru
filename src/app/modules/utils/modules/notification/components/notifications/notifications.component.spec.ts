import {
    async,
    ComponentFixture,
    discardPeriodicTasks,
    fakeAsync,
    TestBed,
    tick,
} from '@angular/core/testing';

import {NotificationsComponent} from './notifications.component';
import {NotificationService} from '../../services/notification.service';
import {instance, mock, when} from 'ts-mockito';
import {Subject} from 'rxjs';
import {NotificationModel} from '../../models/notification';
import {MockModule} from 'ng-mocks';
import {NotificationModule} from '../notification/notification.module';
import {NotificationComponent} from '../notification/notification.component';
import {BrowserDynamicTestingModule} from '@angular/platform-browser-dynamic/testing';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';
import {HttpClientModule} from '@angular/common/http';
import {By} from '@angular/platform-browser';
import {MatIconModule} from '@angular/material/icon';

describe('NotificationsComponent - компонент для показа нотификаций', () => {
    let component: NotificationsComponent;
    let fixture: ComponentFixture<NotificationsComponent>;
    let notificationServiceMock: NotificationService;
    let notificationStream$: Subject<NotificationModel>;

    beforeEach(() => {
        notificationStream$ = new Subject();

        notificationServiceMock = mock(NotificationService);
    });

    beforeEach(() => {
        when(notificationServiceMock.notificationsStream$).thenReturn(
            notificationStream$,
        );
    });

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            declarations: [NotificationsComponent],
            imports: [NotificationModule, NoopAnimationsModule, HttpClientModule],
            providers: [
                {
                    provide: NotificationService,
                    useFactory: () => instance(notificationServiceMock),
                },
            ],
        })
            .overrideModule(BrowserDynamicTestingModule, {
                set: {entryComponents: [NotificationComponent]},
            })
            .overrideModule(NotificationModule, {
                remove: {imports: [MatIconModule]},
                add: {imports: [MockModule(MatIconModule)]},
            })
            .compileComponents();
    }));

    beforeEach(() => {
        fixture = TestBed.createComponent(NotificationsComponent);
        component = fixture.componentInstance;
    });

    it('Если пришла информация о показе нотификации успеха, то она отрисовалась корректно', fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Уведомление',
            type: 'success',
            time: 100,
        });

        fixture.detectChanges();

        // assert
        expect(fixture.nativeElement).toMatchSnapshot();

        tick(100);
        discardPeriodicTasks();
    }));

    it('Если пришла информация о показе нотификации уведомления, то она отрисовалась корректно', fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Внимание',
            type: 'warning',
            time: 100,
        });

        fixture.detectChanges();

        // assert
        expect(fixture.nativeElement).toMatchSnapshot();

        tick(100);
        discardPeriodicTasks();
    }));

    it('Если пришла информация о показе нотификации ошибки, то она отрисовалась корректно', fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Ошибка',
            type: 'error',
            time: 100,
        });

        fixture.detectChanges();

        // assert
        expect(fixture.nativeElement).toMatchSnapshot();

        tick(100);
        discardPeriodicTasks();
    }));

    it('Если пришла информация о показе нотификации, то рисуем нотификацию', fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Уведомление',
            type: 'success',
            time: 100,
        });

        // assert
        expect(fixture.debugElement.query(By.css('app-notification'))).toBeTruthy();

        tick(100);
        discardPeriodicTasks();
    }));

    it(`Если пришла информация о показе нотификации, то сохраняем информацию о ней в объекте`, fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Уведомление',
            type: 'success',
            time: 100,
        });

        // assert
        expect(Object.keys(component.components).length).toBe(1);

        tick(100);
        discardPeriodicTasks();
    }));

    it(`Если пришла информация о показе нотификации и в информации передано время,
        то через указанное время нотификация удалится`, fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Уведомление',
            type: 'success',
            time: 100,
        });

        tick(100);

        // assert
        expect(fixture.debugElement.query(By.css('app-notification'))).toBeNull();

        discardPeriodicTasks();
    }));

    it(`Если пришла информация о показе нотификации и в информации не передано время,
        то через 3 секунды нотификация удалится`, fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Уведомление',
            type: 'success',
        });

        tick(3000);

        // assert
        expect(fixture.debugElement.query(By.css('app-notification'))).toBeNull();

        discardPeriodicTasks();
    }));

    it(`Если пришла информация о показе нотификации и в информации не передано время,
        то через 3 секунды почистится объект с информацией о нотификациях`, fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Уведомление',
            type: 'success',
        });

        tick(3000);

        // assert
        expect(Object.keys(component.components).length).toBe(0);

        discardPeriodicTasks();
    }));

    it(`Если пришла информация о показе нотификации,
        а через секунду пришла новая информация о показе нотификации,
        то на странице будет одновременно находиться 2 нотификации`, fakeAsync(() => {
        // arrange
        fixture.detectChanges();

        // act
        notificationStream$.next({
            text: 'Уведомление',
            type: 'success',
        });

        tick(1000);

        notificationStream$.next({
            text: 'Уведомление 2',
            type: 'warning',
        });

        // assert
        expect(fixture.debugElement.queryAll(By.css('app-notification')).length).toBe(2);

        tick(3000);
        discardPeriodicTasks();
    }));
});
