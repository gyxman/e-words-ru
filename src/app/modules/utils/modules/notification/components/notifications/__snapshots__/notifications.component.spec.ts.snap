// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsComponent - компонент для показа нотификаций Если пришла информация о показе нотификации ошибки, то она отрисовалась корректно 1`] = `
<div
  id="root2"
>
  
  
  <template />
  <app-notification
    class="ng-trigger ng-trigger-expand ng-animating"
    style=""
  >
    
    <div
      class="wrapper wrapper_error"
      style=""
    >
      <div
        class="icon-wrapper"
      >
        <mat-icon
          aria-hidden="true"
          automation-id="icon"
          class="icon"
        />
      </div>
      <div
        class="info"
      >
        Ошибка
      </div>
    </div>
  </app-notification>
</div>
`;

exports[`NotificationsComponent - компонент для показа нотификаций Если пришла информация о показе нотификации уведомления, то она отрисовалась корректно 1`] = `
<div
  id="root1"
>
  
  
  <template />
  <app-notification
    class="ng-trigger ng-trigger-expand ng-animating"
    style=""
  >
    
    <div
      class="wrapper wrapper_warning"
      style=""
    >
      <div
        class="icon-wrapper"
      >
        <mat-icon
          aria-hidden="true"
          automation-id="icon"
          class="icon"
        />
      </div>
      <div
        class="info"
      >
        Внимание
      </div>
    </div>
  </app-notification>
</div>
`;

exports[`NotificationsComponent - компонент для показа нотификаций Если пришла информация о показе нотификации успеха, то она отрисовалась корректно 1`] = `
<div
  id="root0"
>
  
  
  <template />
  <app-notification
    class="ng-trigger ng-trigger-expand ng-animating"
    style=""
  >
    
    <div
      class="wrapper wrapper_success"
      style=""
    >
      <div
        class="icon-wrapper"
      >
        <mat-icon
          aria-hidden="true"
          automation-id="icon"
          class="icon"
        />
      </div>
      <div
        class="info"
      >
        Уведомление
      </div>
    </div>
  </app-notification>
</div>
`;
